### PR TITLE
[SPARK-48416][SQL] Support nested correlated With expression

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

The inner `With` may reference common expressions of an outer `With`. This PR supports this case by making the rule `RewriteWithExpression` only rewrite top-level `With` expressions, and run the rule repeatedly so that the inner `With` expression becomes top-level `With` after one iteration, and gets rewritten in the next iteration.

### Why are the changes needed?

To support optimized filter pushdown with `With` expression.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

updated the unit test

### Was this patch authored or co-authored using generative AI tooling?

no

### CODE REVIEW

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RewriteWithExpressionSuite.scala [hunk 1]
```diff
@@ -104,42 +101,60 @@ class RewriteWithExpressionSuite extends PlanTest {
...
+    val innerExpr2 = With(ref2 + outerRef, Seq(commonExprDef2))
     val outerExpr2 = With(outerRef + innerExpr2, Seq(outerCommonExprDef))
-    intercept[SparkException](Optimizer.execute(testRelation.select(outerExpr2.as("col"))))
+    comparePlans(
+      Optimizer.execute(testRelation.select(outerExpr2.as("col"))),
+      testRelation
+        // The first Project contains the common expression of the outer With
+        .select(star(), rewrittenOuterExpr)
+        // The second Project contains the common expression of the inner With, which does not
+        // references the common expression of the outer With.
```
Comment (zml1206) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RewriteWithExpressionSuite.scala:151]: nit: references -> reference

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RewriteWithExpressionSuite.scala [hunk 2]
```diff
@@ -29,7 +28,7 @@ import org.apache.spark.sql.catalyst.rules.RuleExecutor
 class RewriteWithExpressionSuite extends PlanTest {
 
   object Optimizer extends RuleExecutor[LogicalPlan] {
-    val batches = Batch("Rewrite With expression", Once,
+    val batches = Batch("Rewrite With expression", FixedPoint(5),
```
Comment (MaxGekk) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RewriteWithExpressionSuite.scala:31]: Do you support any levels of nested `with` or there is some restrictions? 5 is enough in any case?

Comment (cloud-fan) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RewriteWithExpressionSuite.scala:31]: The implementation supports any nesting levels but we only test 2 levels here so `FixedPoint(5)` is sufficient

**Review by MaxGekk — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks for the review, merging to master!

**Pull request closed without merging**